### PR TITLE
Add new flow for creating or updating a single provider user

### PR DESCRIPTION
**Pull request by TobyRet**
## Context

As part of the Transition team's provider onboarding process, they need to be able to add users individually and in bulk.

The process of adding users individually has been [updated in the prototype](https://support-for-apply-prototype.herokuapp.com/providers/T92/users
) - this PR focuses solely on the single provider user journey

## Changes proposed in this pull request

- Add feature flag
- Creating a new ProviderUser **is now scoped to a single Provider**. Before it was possible to add permissions for multiple providers
- If that same ProviderUser is created elsewhere, but scoped to a different provider, then we update the ProviderPermissions for that ProviderUser, i.e we do not create a new ProviderUser, rather we add the permissions to the existing ProviderUser.
- When editing a provider user's permissions, we display all Providers that that particular ProviderUser is associated with.

## Guidance to review

These changes may potentially be a little confusing so do let me know if you prefer a demo.

To view the changes:

- the `create_update_single_provider_user` feature flag needs to be switched on.
- the provider must have synced courses on Apply (is this still a requirement)

### Creating a new provider user

creating_a_provider_user

### Add provider permissions to existing provider user

add_permissions_to_existing_user

### Editing a provider user's permissions

editing_permissions

## Link to Trello card

https://trello.com/c/ow63AQnC/3389-%E2%9B%B5%EF%B8%8F-epic-add-provider-user

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/controllers/support_interface/new_provider_users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
+module SupportInterface
+  class NewProviderUsersController < SupportInterfaceController
```
Comment (TobyRet) [inline app/controllers/support_interface/new_provider_users_controller.rb:2]: This a rubbish name but temporary. It is to distinguish this controller from the existing [ProviderUsersController](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/controllers/support_interface/provider_users_controller.rb). Once we want to turn the feature permanently we can rename

File: app/controllers/support_interface/new_provider_users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    def new
+      @provider = Provider.find(provider_id_param)
+      @form = NewProviderUserForm.new(provider_id: provider_id_param)
```
Comment (TobyRet) [inline app/controllers/support_interface/new_provider_users_controller.rb:5]: Same here re: form name
https://github.com/DFE-Digital/apply-for-teacher-training/pull/4828/files#r639814295

File: db/schema.rb [hunk 3]
```diff
@@ -13,6 +13,7 @@
 ActiveRecord::Schema.define(version: 2021_05_25_095046) do
 
   create_sequence "application_choices_id_seq"
+  create_sequence "application_choices_id_seq1"
```
Comment (TobyRet) [inline db/schema.rb:16]: WTF? I'll remove

**Review by cwrw — COMMENTED** [review]
caught up in person, a few preliminary comments, will do a deep dive into the code shortly

File: app/controllers/support_interface/new_provider_users_controller.rb (hunk 1 above)
Comment (TobyRet) [reply app/controllers/support_interface/new_provider_users_controller.rb:2]: 👍

File: app/controllers/support_interface/new_provider_users_controller.rb (hunk 2 above)
Comment (TobyRet) [reply app/controllers/support_interface/new_provider_users_controller.rb:5]: There are two forms for a single provider user - one for creating and another for editing (the latter accommodates all of a provider user's permissions). I've renamed the forms CreateSingleProviderUserForm  and EditSingleProviderUserForm for now. We can revisit once we start the bulk upload

File: app/views/support_interface/providers/users.html.erb [hunk 4]
```diff
@@ -1,6 +1,12 @@
 <%= render 'provider_navigation', title: 'Provider users' %>
 
-<%= govuk_link_to 'Add provider user', new_support_interface_provider_user_path, button: true %>
+<% unless FeatureFlag.active?(:create_update_single_provider_user) %>
+  <%= govuk_link_to 'Add provider user', new_support_interface_provider_user_path, button: true %>
+<% end %>
+
+<% if FeatureFlag.active?(:create_update_single_provider_user) && @provider.sync_courses %>
```
[the remark this answers is not in the document]
Comment (TobyRet) [reply app/views/support_interface/providers/users.html.erb:7]: @simonwhatley - can you confirm whether this is the case or not?

File: app/forms/support_interface/new_provider_user_form.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    attr_accessor :first_name, :last_name, :provider_user, :provider_id
+    attr_reader :provider_permissions, :email_address
+
+    validates :email_address, :first_name, :last_name, presence: true
+    validates :email_address, valid_for_notify: true
+    validates :provider_permissions, presence: true
+
+    def build
+      return unless valid?
```
[the remark this answers is not in the document]
Comment (TobyRet) [reply app/forms/support_interface/new_provider_user_form.rb:14]: Ok no prob - will fix

File: app/forms/support_interface/new_provider_user_form.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
+module SupportInterface
+  class NewProviderUserForm
+    include ActiveModel::Model
+    include ActiveModel::Validations
+
+    attr_accessor :first_name, :last_name, :provider_user, :provider_id
+    attr_reader :provider_permissions, :email_address
+
+    validates :email_address, :first_name, :last_name, presence: true
+    validates :email_address, valid_for_notify: true
```
[the remark this answers is not in the document]
Comment (TobyRet) [reply app/forms/support_interface/new_provider_user_form.rb:10]: @simonwhatley - can you confirm what validation error message copy you would like to display if a duplicate user is created under the same provider?

File: app/services/feature_flag.rb [hunk 7]
```diff
@@ -34,6 +34,7 @@ def feature
     [:expanded_quals_export, 'Rework the Qualifications export to contain all candidate qualifications', 'Malcolm Baig'],
     [:support_user_change_offered_course, 'Allows support users to offer a different course option for an application choice', 'David Gisbey'],
     [:reference_selection, 'Allow candidates to receive multiple references and then select which two are added to their application', 'Malcolm Baig'],
+    [:create_update_single_provider_user, 'New flow for creating or updating a single ProviderUser', 'Toby Retallick'],
```
[the remark this answers is not in the document]
Comment (TobyRet) [reply app/services/feature_flag.rb:37]: 👍

**Comment by simonwhatley** [comment]
@TobyRet here are some thoughts from my review of the PR

- Breadcrumb should read: `Providers > {{ provider name }} > Users` (ie not "Provider users")
- The "Add user" button should be available for all providers whether they have users or not
- The class for the label for first name, last name and email address fields should be `govuk-label--s` not `govuk-label--m`
- The class for the input for first name, last name and email address fields should be `govuk-input govuk-!-width-two-thirds`
- Email address error should read `Enter an email address in the correct format, like name@example.com`
- When a user is added, the support user should be shown the list of users in the given provider, not a long list of provider users
- Success message should read `User {{ full name }} added`
- When navigating to the user's details, they should be `/providers/:providerId/users/:userId` ie within the provider's user section

**Comment by TobyRet** [comment]
> @TobyRet here are some thoughts from my review of the PR
> 
> * Breadcrumb should read: `Providers > {{ provider name }} > Users` (ie not "Provider users")
> * The "Add user" button should be available for all providers whether they have users or not
> * The class for the label for first name, last name and email address fields should be `govuk-label--s` not `govuk-label--m`
> * The class for the input for first name, last name and email address fields should be `govuk-input govuk-!-width-two-thirds`
> * Email address error should read `Enter an email address in the correct format, like name@example.com`
> * When a user is added, the support user should be shown the list of users in the given provider, not a long list of provider users
> * Success message should read `User {{ full name }} added`
> * When navigating to the user's details, they should be `/providers/:providerId/users/:userId` ie within the provider's user section

@simonwhatley - all your recommended changes have been made bar one - _'When navigating to the user's details, they should be `/providers/:providerId/users/:userId` ie within the provider's user section'_

We would need to remove this ProviderUsers route resource and re-define new routes. 
https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/config/routes.rb#L1003

Changing existing routes is a significant change and would perhaps be best dealt with as a separate ticket / PR - would you agree @cwrw ?

File: app/views/support_interface/providers/users.html.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (TobyRet) [reply app/views/support_interface/providers/users.html.erb]: https://github.com/DFE-Digital/apply-for-teacher-training/pull/4828#issuecomment-850342975

File: app/forms/support_interface/new_provider_user_form.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (TobyRet) [reply app/forms/support_interface/new_provider_user_form.rb]: So looks like the issue here was that the ProviderPermissionsForm was using a memoized instance (i.e. the submitted values were not being returned)

https://github.com/DFE-Digital/apply-for-teacher-training/pull/4828/commits/9d99aa879ec70690e7a0e588d8e7d3bce467cdaf

File: app/forms/support_interface/edit_single_provider_user_form.rb [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      else
+        existing_permissions_for_user = []
+      end
+
+      existing_permissions_for_user
+    end
+
+    def possible_permissions
+      @possible_permissions ||= begin
+        if provider_user
```
Comment (cwrw) [inline app/forms/support_interface/edit_single_provider_user_form.rb:37]: maybe we can DRY this up with `all_possible_permissions`?

File: app/forms/support_interface/edit_single_provider_user_form.rb [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    attr_accessor :provider_user, :provider_id
+    attr_reader :provider_permissions
+
+    validates :provider_permissions, presence: true
+
+    def persisted?
+      @provider_user&.persisted?
+    end
+
+    def permission_form
```
Comment (cwrw) [inline app/forms/support_interface/edit_single_provider_user_form.rb:15]: are we using this anywhere?

**Review by cwrw — COMMENTED** [review]
> Changing existing routes is a significant change and would perhaps be best dealt with as a separate ticket / PR - would you agree @cwrw ?

Yes changing existing routes is out of scope for this work. However for the new edit permissions route I think we could include the provider id in `'/providers/users/:provider_user_id'` since we are defining the new permissions edits routes?

I've had a second look and all looks good apart from some minor comments, only thing I think left is when a user exists we don't throw a 500, I guess we're still waiting on the validation error?

**Review by TobyRet — COMMENTED** [review]
> I've had a second look and all looks good apart from some minor comments, only thing I think left is when a user exists we don't throw a 500, I guess we're still waiting on the validation error?

Woops - completely forgot about the dup user - change made here - https://github.com/DFE-Digital/apply-for-teacher-training/pull/4828/commits/143b62da8a7c7ee121c4facf57d8f05464a3a185

**Review by cwrw — APPROVED** [review]
Looks great! Thanks for addressing all comments @TobyRet 💯 

Regarding the routes I've had another think and let's keep as is and aligned and can tackle later when we rip out the older code so all good.

**Pull request merged**